#Exploring Quantum Entanglement and Superposition in Quantum Computing

## Introduction

Quantum entanglement and superposition are fundamental concepts in quantum computing that enable its unique capabilities.

***Quantum Superposition :- *** In classical computing, a bit can be either 0 or 1. However, in quantum computing, a quantum bit (qubit) can exist in a state of superposition, meaning it can be both 0 and 1 simultaneously. This is due to the principles of quantum mechanics, where particles can exist in multiple states at once. Superposition allows quantum computers to process a vast amount of information simultaneously, providing a significant speedup for certain computations.

**Quantum Entanglement :-** Quantum entanglement is a phenomenon where two or more qubits become interconnected such that the state of one qubit directly influences the state of the other, no matter the distance between them. This correlation is stronger than any classical correlation and is a key resource for quantum computing. Entanglement enables quantum computers to perform complex calculations more efficiently by leveraging these strong correlations.


**Role of Quantum Computing in Data**

1.   Data Processing: Superposition allows quantum computers to handle and process large datasets more efficiently than classical computers. For example, quantum algorithms like Grover’s search algorithm can search unsorted databases exponentially faster than classical algorithms.
2.   Data Security: Entanglement is used in quantum cryptography to create secure communication channels. Quantum key distribution (QKD) protocols, such as BB84, use entangled particles to detect eavesdropping, ensuring data security.
1.   Machine Learning: Quantum machine learning algorithms can potentially handle high-dimensional data more effectively.





## Necessary Libraries/Modules

In [ ]:
!pip install qiskit
!pip install qiskit-aer

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.7/4.7 MB 45.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 8.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 59.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.7/49.7 kB 3.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 39.4/39.4 MB 16.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 108.5/108.5 kB 7.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.3/12.3 MB 85.4 MB/s eta 0:00:00


**Note :-** Qiskit is an open-source framework for quantum computing that allows users to design, simulate, and execute quantum algorithms on quantum hardware and simulators.

In [ ]:
import qiskit
print(qiskit.__version__)

1.2.1


In [ ]:
from qiskit import QuantumCircuit
from qiskit import assemble
from qiskit_aer import Aer
import pandas as pd
import matplotlib.pyplot as plt
import plotly.express as px

## Quantum circuit to explain superposition

In [ ]:
# Create a Quantum Circuit with 1 qubit and 1 classical bit
qc = QuantumCircuit(1, 1)

# Apply Hadamard gate to the qubit
qc.h(0)

# Measure the qubit
qc.measure(0, 0)

# Print the quantum circuit
print(qc)

     ┌───┐┌─┐
  q: ┤ H ├┤M├
     └───┘└╥┘
c: 1/══════╩═
           0 


*   **Explanation :-**
1.   The qubit starts in the ∣0⟩ state, representing a definite state of 0.
2.   Applying the Hadamard gate creates a superposition of ∣0⟩ and ∣1⟩.
1.   The qubit now has a 50% probability of being measured as 0 and 50% as 1.
2.   When the qubit is measured, the superposition collapses to a definite state either  ∣0⟩ or ∣1⟩.
1.   Due to the superposition, the result is probabilistic, with a 50% chance of each outcome.









### Running the circuit on quantum simulator

In [ ]:
# Get the 'qasm_simulator' backend from the Aer provider
simulator = Aer.get_backend('qasm_simulator')

# Run the quantum circuit 'qc' on the simulator with 1000 shots and get the result
result_sim = simulator.run(qc, shots=1000).result()

# Extract the counts of each outcome from the simulation result
counts_sim = result_sim.get_counts(qc)
print("Simulation Result:")
print(counts_sim)

# Convert the counts dictionary into a DataFrame for easier visualization
df_counts = pd.DataFrame(list(counts_sim.items()), columns=['Outcome', 'Counts'])

# Create and display a bar chart of the simulation results using Plotly Express
px.bar(df_counts, x='Outcome', y='Counts', title='Bar Graph of Simulation Results')

Simulation Result:
{'0': 512, '1': 488}




*   **Insights :-**

1.   The Hadamard gate created a superposition, resulting in nearly equal probabilities for ∣0⟩ and ∣1⟩.







## Quantum circuit to explain entanglement

In [ ]:
# Create a Quantum Circuit with 2 qubits and 2 classical bits
qc = QuantumCircuit(2, 2)

# Apply Hadamard gate to the first qubit
qc.h(0)

# Apply CNOT gate to entangle the qubits
qc.cx(0, 1) # Use cx() instead of cnot()

# Measure both qubits
qc.measure(0, 0)
qc.measure(1, 1)

# Print the quantum circuit
print(qc)

     ┌───┐     ┌─┐   
q_0: ┤ H ├──■──┤M├───
     └───┘┌─┴─┐└╥┘┌─┐
q_1: ─────┤ X ├─╫─┤M├
          └───┘ ║ └╥┘
c: 2/═══════════╩══╩═
                0  1 




*   **Explanation :-**
1.   The Hadamard gate puts qubit 0 into a superposition of ∣0⟩ and ∣1⟩.
2.   The CNOT gate entangles qubit 0 with qubit 1, such that qubit 1’s state is dependent on qubit 0’s state.
1.   The circuit measures both qubits. Due to the entanglement, the measurement outcomes of qubits 0 and 1 are correlated. If qubit 0 is measured as 0, qubit 1 will also be measured as 0, and if qubit 0 is measured as 1, qubit 1 will be measured as 1.






### Running the circuit on quantum simulator

In [ ]:
# Get the 'qasm_simulator' backend from the Aer provider
simulator = Aer.get_backend('qasm_simulator')

# Run the quantum circuit 'qc' on the simulator with 1000 shots and get the result
result_sim = simulator.run(qc, shots=1000).result()

# Extract the counts of each outcome from the simulation result
counts_sim = result_sim.get_counts(qc)
print("Simulation Result:")
print(counts_sim)

# Convert the counts dictionary into a DataFrame for easier visualization
df_counts = pd.DataFrame(list(counts_sim.items()), columns=['Outcome', 'Counts'])

# Create a bar chart of the simulation results using Plotly Express
fig = px.bar(df_counts, x='Outcome', y='Counts', title='Bar Graph of Simulation Results')

# Display the bar chart
fig.show()

Simulation Result:
{'00': 492, '11': 508}




*   **Insights :-**

1.   The qubit measurement resulted in both qubits being 0 in 492 out of 1000 trials.
2.   The qubit measurement resulted in both qubits being 1 in 508 out of 1000 trials.
1.   The result shows that when qubit 0 was measured as 0, qubit 1 was also measured as 0. Similarly, when qubit 0 was measured as 1, qubit 1 was measured as 1. This confirms that the qubits are entangled. Their measurement results are correlated.







## Trying the quantum circuit on real quantum computer

**Note :-** To run the previous circuit simulations on real quantum computers, first register on the IBM Quantum platform and obtain an API token. Due to potential long queues for accessing quantum hardware, execute the below code for each gate one at a time. Be prepared for possible wait times ranging from several minutes to several hours to access the quantum computers.

In [ ]:
!pip install qiskit-ibm-runtime

In [ ]:
from qiskit_ibm_runtime import QiskitRuntimeService

service = QiskitRuntimeService(channel="ibm_quantum", token="API Token")

In [ ]:
from qiskit import QuantumCircuit, transpile
from qiskit_ibm_runtime import QiskitRuntimeService, SamplerV2 as Sampler

# Choose the least busy real quantum backend
backend = service.least_busy(operational=True, simulator=False)

# Create a sampler for running the circuit
sampler = Sampler(backend)

# Transpile the circuit to the backend's native gate set
qc_transpiled = transpile(qc, backend)

# Submit the job to run the transpiled circuit
job = sampler.run([qc_transpiled])
print(f"Job ID: {job.job_id()}")

# Get and print the result
result = job.result()
print(result)

Job ID: cvj7v3c8w2g0008esc7g
